llvm/include/llvm/IR/PassManager.h

```diff
@@ -1317,6 +1351,23 @@ struct InvalidateAnalysisPass
     auto PassName = MapClassName2PassName(ClassName);
     OS << "invalidate<" << PassName << '>';
   }
+
+  // MachineFunctionPass interface, define them separately to prevent
+  // dependency on CodeGen
+  template <typename MachineFunctionPropertiesT = MachineFunctionProperties>
+  static MachineFunctionPropertiesT getRequiredProperties() {
+    return MachineFunctionPropertiesT();
+  }
+
+  template <typename MachineFunctionPropertiesT = MachineFunctionProperties>
```

Comment:
oh, are these templates just to work around the IR -> CodeGen depenency? that's not great

let me try some things locally, like making a completely separate MachineFunctionPassManager

Reply:
oh I see that `PassModel` in PassManagerInternal.h also does this...